src/engine/Server.cpp

```diff
@@ -335,47 +335,39 @@ string Server::composeResponseJson(const ParsedQuery& query,
   off_t compResultUsecs = _requestProcessingTimer.usecs();
   size_t resultSize = rt->size();
 
-  std::ostringstream os;
-  os << "{\n"
-     << "\"query\": " << ad_utility::toJson(query._originalString) << ",\n"
-     << "\"status\": \"OK\",\n"
-     << "\"resultsize\": \"" << resultSize << "\",\n";
-
-  os << "\"selected\": ";
-  if (query._selectedVariables.size()) {
-    os << "[\"" << ad_utility::join(query._selectedVariables, "\", \"")
-       << "\"],\n";
-  } else {
-    os << "[],\n";
-  }
+  nlohmann::json j;
 
-  os << "\"runtimeInformation\" : ";
-  os << RuntimeInformation::ordered_json(
+  j["query"] = query._originalString;
+  j["status"] = "OK";
+  j["resultsize"] = resultSize;
+  j["warnings"] = qet.collectWarnings();
+  j["selected"] = query._selectedVariables;
+
+  j["runtimeInformation"] = RuntimeInformation::ordered_json(
       qet.getRootOperation()->getRuntimeInfo());
-  os << ", \n";
 
-  os << "\"res\": ";
-  size_t limit = MAX_NOF_ROWS_IN_RESULT;
-  size_t offset = 0;
-  if (query._limit.size() > 0) {
-    limit = static_cast<size_t>(atol(query._limit.c_str()));
-  }
-  if (query._offset.size() > 0) {
-    offset = static_cast<size_t>(atol(query._offset.c_str()));
+  {
+    std::stringstream os;
+    size_t limit = MAX_NOF_ROWS_IN_RESULT;
+    size_t offset = 0;
+    if (query._limit.size() > 0) {
+      limit = static_cast<size_t>(atol(query._limit.c_str()));
+    }
+    if (query._offset.size() > 0) {
+      offset = static_cast<size_t>(atol(query._offset.c_str()));
+    }
+    _requestProcessingTimer.cont();
+    qet.writeResultToStreamAsJson(os, query._selectedVariables, limit, offset,
+                                  maxSend);
+    j["res"] = json::parse(os.str());
```

Comment:
Is this efficient for a huge result (e.g. millions or billions of triples)?

However, the UI only asks for the Top-X results and big results are usually exported as CSV or TSV